tuist/tuist
# Optimize `BuildPhaseGenerator`'s method of detecting assets and localized files

Part of: https://github.com/tuist/tuist/issues/1093

### Short description 📝

While profiling the workspace generator (#1093) the regex checking of assets and localised file paths stood out as one of the areas that could be simplified to gain some performance improvements.

### Solution 📦

- Leverage simpler string contains check to determine if a path is nested within an `xcassets` or `lproj` container
- The check on the `pathString` can be done directly as `AbsolutePath` ensures no trailing `/` in its `pathString` as such checking for `.xcassets/` or `.lproj/` can act as a suitable faster replacement of the `regex` check that was previously used.


### Implementation 👩‍💻👨‍💻

- [x] Update `BuildPhaseGenerator`

### Test Plan 🛠

- Run `tuist generate` within `fixtures/ios_app_with_framework_and_resources`
- Verify the generated projects contains the xcassets and localized string files as before

### Profiling Results ⏱

From profiling #1093, this change can improve the `BuildPhaseGenerator ` time  by ~90% 


| Component                  | Duration   | Percentage | Delta     |
| -------------------------- | ---------- | ---------- | --------- |
| WorkspaceGenerator         |  39.50 s   |  100.00%   | -8.29%    |
| ProjectGenerator           |  39.49 s   |  99.97%    | -8.29%    |
| ProjectFileElements        |  20.91 s   |  52.94%    | -1.65%    |
| TargetGenerator            |  18.06 s   |  45.72%    | -15.13%   |
| LinkGenerator              |  17.62 s   |  44.61%    | -1.34%    |
| BuildPhaseGenerator        |  340.17 ms |  0.86%     | -89.72%   |
| ConfigGenerator            |  89.55 ms  |  0.23%     | -3.12%    |
